bundles/org.openhab.core/src/main/java/org/openhab/core/internal/auth/UserRegistryImpl.java

```diff
@@ -128,23 +133,132 @@ public User register(String username, String password, Set<String> roles) {
+            }
+
+            return new Authentication(managedUser.getName(), managedUser.getRoles().stream().toArray(String[]::new));
+        } else if (credentials instanceof UserApiTokenCredentials) {
+            UserApiTokenCredentials apiTokenCreds = (UserApiTokenCredentials) credentials;
+            String[] apiTokenParts = apiTokenCreds.getApiToken().split("\\.");
+            if (apiTokenParts.length != 3 || !APITOKEN_PREFIX.equals(apiTokenParts[0])) {
+                throw new AuthenticationException("Invalid API token format");
+            }
+            for (User user : getAll()) {
+                ManagedUser managedUser = (ManagedUser) user;
+                for (UserApiToken userApiToken : managedUser.getApiTokens()) {
+                    // only check if the name in the token matches
+                    if (!userApiToken.getName().equals(apiTokenParts[1])) {
+                        continue;
+                    }
+                    String[] existingTokenHashAndSalt = userApiToken.getApiToken().split(":");
+                    String incomingTokenHash = hash(apiTokenCreds.getApiToken(), existingTokenHashAndSalt[1],
+                            APITOKEN_ITERATIONS).get();
+
+                    if (incomingTokenHash.equals(existingTokenHashAndSalt[0])) {
+                        return new Authentication(managedUser.getName(),
+                                managedUser.getRoles().stream().toArray(String[]::new), userApiToken.getScope());
+                    }
+                }
+            }
+
+            throw new AuthenticationException("Unknown API token");
+        }
+
+        throw new IllegalArgumentException("Invalid credential type");
+    }
+
+    @Override
+    public void changePassword(User user, String newPassword) {
+        if (!(user instanceof ManagedUser)) {
+            throw new IllegalArgumentException("User is not managed: " + user.getName());
+        }
+
+        ManagedUser managedUser = (ManagedUser) user;
+        String passwordSalt = generateSalt(KEY_LENGTH / 8).get();
+        String passwordHash = hash(newPassword, passwordSalt, PASSWORD_ITERATIONS).get();
+        managedUser.setPasswordSalt(passwordSalt);
+        managedUser.setPasswordHash(passwordHash);
+        update(user);
+    }
+
+    @Override
+    public void addUserSession(User user, UserSession session) {
+        if (!(user instanceof ManagedUser)) {
+            throw new IllegalArgumentException("User is not managed: " + user.getName());
+        }
+
+        ManagedUser managedUser = (ManagedUser) user;
+        managedUser.getSessions().add(session);
+        update(user);
+    }
+
+    @Override
+    public void removeUserSession(User user, UserSession session) {
+        if (!(user instanceof ManagedUser)) {
+            throw new IllegalArgumentException("User is not managed: " + user.getName());
         }
+
+        ManagedUser managedUser = (ManagedUser) user;
+        managedUser.getSessions().add(session);
```

Comment:
That's probably because I am not very careful :) 🙄

Reply:
(and, admittedly, not too big on unit tests...)

Reply:
Yes that explains!

Reply:
I also noticed the test coverage is lacking on such critical code. But I'm OK with that if you promise to make up for that in a follow up PR. :wink:

Reply:
No time like the present, I added some test coverage to the UserRegistryImpl :)

Reply:
That's even better. :+1: I sometimes run out of time myself too while very enthusiastically implementing lots of new functionalities. :smile: